R/PipeOpDecode.R

```diff
@@ -0,0 +1,240 @@
+#'   is set to `"ref"` in rows where the value is less than or equal to a specified cutoff value (e.g., `0`) in all
+#'   columns in that group To change the name of the reference level, use [`PipeOp???`] (Mutate? ColApply?).
+#'   Initialized to `0`.
+#' * `ties_method` :: `character(1)`\cr
+#'   Method for resolving ties if multiple columns have the same value. Specifies the value from which of the columns
+#'   with the same value is to be picked. Options are `"first"`, `"last"`, or `"random"`. Initialized to `"random"`.
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpTaskPreprocSimple`]/[`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @family PipeOps
+#' @template seealso_pipeopslist
+#' @include PipeOpTaskPreproc.R
+#' @export
+#' @examples
+#' library("mlr3")
+#'
+#' # Create example task with one-hot encoding
+#' df = data.frame(
+#'   target = runif(10),
+#'   x.1 = rep(c(1, 0), 5),
+#'   x.2 = rep(c(0, 1), 5),
+#'   y.1 = rep(c(1, 0), 5),
+#'   y.2 = rep(c(0, 1), 5),
+#'   a = runif(10)
+#' )
+#' task = TaskRegr$new(id = "example", backend = df, target = "target")
+#'
+#' pop = po("decode")
+#'
+#' # Training
+#' train_out = pop$train(list(task))[[1]]
+#' # x.1 and x.2 are collapsed into x, same for y; a is ignored.
+#' train_out$data()
+#'
+#' # Create example task with treatment encoding
+#' df = data.frame(
+#'   target = runif(15),
+#'   x.1 = rep(c(1, 0, 0), 5),
+#'   x.2 = rep(c(0, 1, 0), 5)
+#' )
+#' task = TaskRegr$new(id = "example", backend = df, target = "target")
+#'
+#' pop = po("decode")
+#' pop$param_set$set_values(treatment_encoding = TRUE)
+#'
+#' # Training
+#' train_out = pop$train(list(task))[[1]]
+#' # x.1 and x.2 are collapsed into x; in rows where all values
+#' # are smaller or equal to 0, the reference level is set
+#' train_out$data()
+#'
+PipeOpDecode = R6Class("PipeOpDecode",
+  inherit = PipeOpTaskPreprocSimple,
+  public = list(
+    initialize = function(id = "decode", param_vals = list()) {
+      ps = ps(
+        group_pattern = p_uty(custom_check = check_string, tags = c("train", "required")),
+        treatment_encoding = p_lgl(tags = c("train", "required")),
+        treatment_cutoff = p_dbl(default = 0, tags = "train", depends = quote(treatment_encoding == TRUE)),
+        ties_method = p_fct(c("first", "last", "random"), tags = c("train", "required"))
+      )
+      ps$values = list(treatment_encoding = FALSE, group_pattern = "^([^.]+)\\.", ties_method = "random")
+      super$initialize(id, param_set = ps, param_vals = param_vals, tags = "encode", feature_types = c("integer", "numeric"))
+    }
+  ),
+  private = list(
+
+    .get_state_dt = function(dt, levels, target) {
+      pv = self$param_set$values
+      cols = colnames(dt)
+
+      if (pv$treatment_encoding) {
+        # Determine name for reference level
+        ref_name = "ref"
+        counter = 1
+        while (ref_name %in% cols) {
```

Comment:
This should be a function that is then called at each occurrence of `... = ref_name`, since the collision check needs to be done for each individual group separately.